File: backend/src/modules/categories/categories.repository.ts
import { PrismaClient } from '@prisma/client';
import { prisma } from '../../infra/database/prisma.client';

export class CategoriesRepository {
  constructor(private readonly db: PrismaClient = prisma) {}

  findAll(onlyActive: boolean) {
    return this.db.category.findMany({
      where: onlyActive ? { active: true } : undefined,
      orderBy: { name: 'asc' }
    });
  }

  findById(id: string) {
    return this.db.category.findUnique({ where: { id } });
  }

  findByName(name: string) {
    return this.db.category.findUnique({ where: { name } });
  }

  create(data: { name: string; icon?: string; color?: string }) {
    return this.db.category.create({ data });
  }

  update(id: string, data: { name?: string; icon?: string; color?: string; active?: boolean }) {
    return this.db.category.update({ where: { id }, data });
  }
}

File: desktop/lib/features/categories/domain/repositories/categories_repository.dart
import '../../../../core/error/result.dart';
import '../entities/category.dart';

abstract class CategoriesRepository {
  Future<Result<List<Category>>> list();
  Future<Result<Category>> create({required String name, String? icon, String? color});
  Future<Result<Category>> update({
    required String id,
    String? name,
    String? icon,
    String? color,
    bool? active,
  });
}
